### PR TITLE
Check whether the parcel's signer matches the asset's registrar or not

### PR DESCRIPTION
**Pull request by majecty**

### CODE REVIEW

**Comment by majecty** [comment]
I'm adding integration tests to the SDK.
I'll remove "[WIP]" after adding the tests.

File: state/src/impls/shard_level.rs [hunk 1]
```diff
@@ -222,11 +222,30 @@ impl<B: Backend + ShardBackend> ShardLevelState<B> {
...
+            let asset_scheme_address = AssetSchemeAddress::from_hash(asset_type)
+                .ok_or(TransactionError::AssetSchemeNotFound(asset_type.into()))?;
+            let asset_scheme = self
+                .asset_scheme((&asset_scheme_address).into())?
+                .ok_or(TransactionError::AssetSchemeNotFound(asset_scheme_address.into()))?;
+
+            if let Some(ref registrar) = asset_scheme.registrar() {
+                if registrar != sender {
+                    let mismatch = Mismatch {
+                        expected: registrar.clone(),
```
Comment (sgkim126) [inline state/src/impls/shard_level.rs]:
```rust
                        expected: *registrar,
                        found: *sender,
```
is better in general.

File: state/src/impls/shard_level.rs [hunk 2]
```diff
@@ -222,11 +222,30 @@ impl<B: Backend + ShardBackend> ShardLevelState<B> {
...
+            let asset_type = input.prev_out.asset_type.clone();
+            let asset_scheme_address = AssetSchemeAddress::from_hash(asset_type)
+                .ok_or(TransactionError::AssetSchemeNotFound(asset_type.into()))?;
+            let asset_scheme = self
+                .asset_scheme((&asset_scheme_address).into())?
+                .ok_or(TransactionError::AssetSchemeNotFound(asset_scheme_address.into()))?;
+
+            if let Some(ref registrar) = asset_scheme.registrar() {
+                if registrar != sender {
+                    let mismatch = Mismatch {
```
Comment (kseo) [inline state/src/impls/shard_level.rs]: Do we need an intermediate local variable here? Let's inline it.

File: types/src/transaction/error.rs [hunk 3]
```diff
@@ -55,6 +55,7 @@ pub enum Error {
     InvalidWorldId(WorldId),
     InvalidWorldNonce(Mismatch<u64>),
     EmptyShardOwners(ShardId),
+    RegistrarNotMatched(Mismatch<Address>),
```
Comment (kseo) [inline types/src/transaction/error.rs]: `RegistrarNotMatched` -> `NotRegistrar`

**Comment by majecty** [comment]
@sgkim126 @kseo I fixed your requests.

**Comment by sgkim126** [comment]
LGTM

File: types/src/transaction/error.rs [hunk 4]
```diff
@@ -63,6 +64,7 @@ const ERROR_ID_INSUFFICIENT_PERMISSION: u8 = 13u8;
 const ERROR_ID_INVALID_WORLD_ID: u8 = 14u8;
 const ERROR_ID_INVALID_WORLD_NONCE: u8 = 15u8;
 const ERROR_ID_EMPTY_SHARD_OWNERS: u8 = 16u8;
+const ERROR_ID_REGISTRAR_NOT_MATCHED: u8 = 17u8;
```
Comment (kseo) [inline types/src/transaction/error.rs]: `ERROR_ID_REGISTRAR_NOT_MATCHED` -> `ERROR_ID_NOT_REGISTRAR`

File: types/src/transaction/error.rs [hunk 5]
```diff
@@ -173,6 +182,11 @@ impl Display for Error {
             Error::InvalidWorldId(_) => write!(f, "The world id is invalid"),
             Error::InvalidWorldNonce(mismatch) => write!(f, "The world nonce {}", mismatch),
             Error::EmptyShardOwners(shard_id) => write!(f, "Shard({}) must have at least one owner", shard_id),
+            Error::NotRegistrar(mismatch) => write!(
+                f,
+                "The signer of the parcel({}) is not matched with the asset's registrar({})",
```
Comment (kseo) [inline types/src/transaction/error.rs]: does not match

**Pull request merged**